### PR TITLE
[node] Fix `http.createServer` signature

### PR DESCRIPTION
**Pull request by Lordfirespeed**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

I am changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: [none applicable]

Basically, `createServer`'s current definition doesn't work when using custom prototypes like the following for `IncomingMessage`, `ServerResponse`:

```ts
class MyRequest extends http.IncomingMessage {
    foo: string 
}

class MyResponse<Request extends MyRequest = MyRequest> extends http.ServerResponse<Request> {
    foo: string
}

createServer({ IncomingMessage: MyRequest, ServerResponse: MyResponse }) // this fails
```

because the construct signature of `MyResponse` doesn't match 
```ts
new <Request extends IncomingMessage = IncomingMessage>(request: Request): ServerResponse<Request>
```
Which is the wrong signature to match against, it should be matching against `ServerResponse<InstanceType<Request>>` instead of `ServerResponse`, which is equivalent to `ServerResponse<IncomingMessage>`

`https.d.ts` and `http2.d.ts` have the same/similar issues, ~~I haven't addressed those in this PR~~ I have now addressed those in this PR.

### CODE REVIEW

**Comment by Lordfirespeed** [comment]
I don't understand how to fix the build. The errors seem unrelated to my changes?

**Comment by Lordfirespeed** [comment]
The build is failing now because of a breaking release to `@npm/types` [here](https://github.com/npm/types/releases/tag/v2.0.0)

**Comment by Lordfirespeed** [comment]
The build is failing because ... no good reason ? 😩

**Comment by forivall** [comment]
CI is showing the following:
```
Error in fontagon
Error: 
/home/runner/work/DefinitelyTyped/DefinitelyTyped/types/fontagon/index.d.ts
  2:38  error  TypeScript@4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5, 5.6 compile error: 
Cannot find module 'svgicons2svgfont' or its corresponding type declarations  @definitelytyped/expect

✖ 1 problem (1 error, 0 warnings)
```

**Comment by Lordfirespeed** [comment]
Yes, and what's that got to do with this PR?!

**Comment by nicolas377** [comment]
I've been summoned from my summer break grave, and oh boy did I just go on the tour of this repo. Looks like the original reason my package is breaking is a semver major release from `@npm/types`, only a problem because of my careless version restrictions. I'm perfectly fine with this PR fixing my package, a maintainer will get around to it soon, sorry! In regards to the failing CI, Fontagon is tested since it's a dep of a dep (ttf2woff references node types), but it was broken because of the recently merged #70261. IMO this is pretty much a failure of the CI on that PR, maintainers was that intentional? The `fontagon` types need to be updated to rely on `svgicons2svgfont` itself (or an older version of the types) since it provides types (pinging @peterblazejewicz since it's late and I'm lazy, sorry). Also, I haven't been around for a sec, are we even supposed to test 2 layer deps? I'm going to bed, but signing off on the change.

**Review by nicolas377 — APPROVED** [review]

**Comment by jakebailey** [comment]
>  IMO this is pretty much a failure of the CI on that PR, maintainers was that intentional? 

This is a known hole in CI; the package isn't shimmed on npm before the PR is merged, of course, so we can't test packages in the repo with it.

**Comment by jakebailey** [comment]
> Yes, and what's that got to do with this PR?!
> 
> I haven't touched that module at all! 😭

A change to the node types can break deps of deps of deps of ... and so on, so the entire transitive tree is checked. Unfortunately, that leaves popular packages sensitive to ecosystem-wide breaks until they are fixed.

**Comment by jakebailey** [comment]
Sent https://github.com/DefinitelyTyped/DefinitelyTyped/pull/70297

**Comment by Lordfirespeed** [comment]
Are there any actions I need to take here, other than waiting for a review?

**Comment by jakebailey** [comment]
It needs to be reviewed, yes, or time out and move into needing a maintainer to look at it manually. I would suggest merging from `master` to pull in the changes to the npm packages which has already been fixed.

**Comment by jakebailey** [comment]
That is a new 5.7 break we are going to fix on master (sorry...)

**Comment by Lordfirespeed** [comment]
> That is a new 5.7 break we are going to fix on master (sorry...)

hahaha, don't worry about it :joy:
I was mostly worried I might have to resolve it myself to keep this PR alive!

It's a case of replacing:
```ts
let BlogPost: typeof Nodal.Model;
```
with
```ts
let BlogPost = undefined as typeof Nodal.Model;
```
right?

**Comment by nicolas377** [comment]
Thanks @jakebailey for fixing the ecosystem breaks, really appreciate it. I'm gonna recuse myself since this doesn't affect any of my packages anymore, but I'll keep the review up out of convenience for myself and support for the changes. Thanks to all for your work!

**Comment by DangerBotOSS** [comment]
## Formatting

The following files are not formatted:
1. types/node/https.d.ts
1. types/node-red__registry/node-red__registry-tests.ts
1. types/node/http2.d.ts
1. types/node/test/http2.ts
1. types/node/v18/https.d.ts

<details>
<summary>as well as these 2 other files...</summary>
<p>types/node/v16/https.d.ts, types/node/v20/https.d.ts</p>
</details>


Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 82ecf3c6931d6962a2dd2baeca3abe2ee184a6b2
</p>

**Review by mcollina — APPROVED** [review]
lgtm

**Review by andrewbranch — APPROVED** [review]

**Comment by Lordfirespeed** [comment]
Ready to merge

**Comment by RobinTail** [comment]
breaks for `express` as well, same error

**Comment by Lordfirespeed** [comment]
> please revert

Reverting isn't the play here - this PR makes the types more correct. We should look into fixing the issue properly. 

I haven't had a chance to look today because I was supposed to be flying home. My flight was cancelled and there are no hotel rooms available in this city. I have just sat down to eat my first meal in about 10 hours. I'm sorry I can't look at this right away.

**Comment by Lordfirespeed** [comment]
What confuses me is that some typescript versions correctly recognise the type as `Server<X, Y>` but others expand out the response type into some huge object type `Server<X, { ... }>`? 

Please can someone try changing the `events` import on line 46 of `http.d.ts` to a namespace import (`import * as events from "events"`) ?

**Comment by andrewbranch** [comment]
Minimal repro without other libraries:

```ts
import https from 'node:https';

export const createServer = () => {
  return https.createServer();
}
```

(The failure in express-zod-api is unrelated to express)

**Comment by Lordfirespeed** [comment]
> Minimal repro without other libraries

What about `http.createServer` and `http2.createServer`? Is the problem localised to `https`?

**Comment by andrewbranch** [comment]
> As a band-aid, though, please can someone try changing the events import on line 46 of http.d.ts to a namespace import

No luck. Exporting two more private interfaces from events.d.ts fixes the error, but I think the indirection of `Request extends typeof http.IncomingMessage` and `InstanceType<Request>` is always going to have a really nasty effect on declaration emit (and quite possibly performance, too). Before:

```ts
import https from 'node:https';
export declare const createServer: () => https.Server<typeof import("http").IncomingMessage, typeof import("http").ServerResponse>;
```

After:

```ts
import https from 'node:https';
export declare const createServer: () => https.Server<typeof import("http").IncomingMessage, {
    new (req: import("http").IncomingMessage): import("http").ServerResponse<import("http").IncomingMessage>;
    fromWeb(writableStream: import("stream/web").WritableStream, options?: Pick<import("stream").WritableOptions, "decodeStrings" | "highWaterMark" | "objectMode" | "signal">): import("stream").Writable;
    toWeb(streamWritable: import("stream").Writable): import("stream/web").WritableStream;
    addAbortSignal<T extends import("stream").Stream>(signal: AbortSignal, stream: T): T;
    getDefaultHighWaterMark(objectMode: boolean): number;
    setDefaultHighWaterMark(objectMode: boolean, value: number): void;
    finished: typeof import("stream").finished;
    pipeline: typeof import("stream").pipeline;
    isErrored(stream: import("stream").Readable | import("stream").Writable | NodeJS.ReadableStream | NodeJS.WritableStream): boolean;
    isReadable(stream: import("stream").Readable | NodeJS.ReadableStream): boolean;
    Stream: typeof import("stream").Stream;
    Readable: typeof import("stream").Readable;
    Writable: typeof import("stream").Writable;
    Duplex: typeof import("stream").Duplex;
    Transform: typeof import("stream").Transform;
    PassThrough: typeof import("stream").PassThrough;
    readonly promises: typeof import("node:stream/promises");
    readonly consumers: typeof import("node:stream/consumers");
    once(emitter: NodeJS.EventEmitter, eventName: string | symbol, options?: import("events").StaticEventEmitterOptions): Promise<any[]>;
    once(emitter: EventTarget, eventName: string, options?: import("events").StaticEventEmitterOptions): Promise<any[]>;
    on(emitter: NodeJS.EventEmitter, eventName: string | symbol, options?: import("events").StaticEventEmitterIteratorOptions): AsyncIterableIterator<any[]>;
    on(emitter: EventTarget, eventName: string, options?: import("events").StaticEventEmitterIteratorOptions): AsyncIterableIterator<any[]>;
    listenerCount(emitter: NodeJS.EventEmitter, eventName: string | symbol): number;
    getEventListeners(emitter: EventTarget | NodeJS.EventEmitter, name: string | symbol): Function[];
    getMaxListeners(emitter: EventTarget | NodeJS.EventEmitter): number;
    setMaxListeners(n?: number, ...eventTargets: Array<EventTarget | NodeJS.EventEmitter>): void;
    addAbortListener(signal: AbortSignal, resource: (event: Event) => void): Disposable;
    readonly errorMonitor: typeof import("events").errorMonitor;
    readonly captureRejectionSymbol: typeof import("events").captureRejectionSymbol;
    captureRejections: boolean;
    defaultMaxListeners: number;
    EventEmitter: typeof import("events");
    EventEmitterAsyncResource: typeof import("events").EventEmitterAsyncResource;
}>;
```

The type parameters really ought to be refactored such that `InstanceType` isn’t needed, but that might be unacceptably breaky for @types/node.

I’m going to revert, sorry. We can’t wait days or even hours when @types/node is broken.

**Comment by Lordfirespeed** [comment]
What's odd is that typescript 5.0 through 5.3 (?) don't do that - I wonder if there's been a regression on typescript's side ...

> I’m going to revert, sorry. We can’t wait days or even hours when @types/node is broken.

Totally understandable. Hopefully I can resolve this issue and bring it back for a second go around without the problems

**Comment by andrewbranch** [comment]
> typescript versions 5.0 through 5.3 (?) don't do the big messy object type

They do, in declaration emit. I’m seeing it in quick info too, but it’s possible that some specific input code on some specific TS version was producing a clean-looking but invalid quick info response. Quick info isn’t always valid code.

**Comment by Lordfirespeed** [comment]
> They do, in declaration emit. [...] Quick info isn’t always valid code.
Thankyou for the insight - that's good to know.

**Comment by Lordfirespeed** [comment]
So @andrewbranch - what if I add overload declarations for the old signature that take precedence?

This way the common use-case of having no custom prototypes should continue to work with no performance issues, but people using custom prototypes could do so unhindered.

Of course I would need to add those `events` imports too.

**Comment by nicolas377** [comment]
> @nicolas377 please revert

In the wise words of someone smart once: "not it"
If y'all actually need me to revert the approval, email me. I don't check github very often (this is my first time checking since I was pinged 3 weeks ago). I'm not gonna be helpful with fixing the problems than anyone else would.

**Pull request merged**
